### PR TITLE
fix(compare.go): add unit tests

### LINKED ISSUE #650: feat: Add Unit Tests for compare.go
#### **Problem**

The OLake codebase currently has unit tests for:

* `resolver`
* `flattener`
* `catalog`

However, the file `olake/utils/typeutils/compare.go` does not have any test coverage.

This file handles critical comparison logic for various data types:

* `Compare()` - Generic comparison function that returns 0 for equal, -1 if a < b, else 1 if a > b
  - Handles nil value comparisons
  - Supports unsigned integer
  - Supports signed integer
  - Supports floating point number
  - Supports time.Time comparisons
  - Supports boolean comparisons
  - Supports custom Time type comparisons
  - Falls back to string comparison for other types

Without unit tests, bugs in type comparisons, edge cases with nil values, or incorrect comparison results may slip through during development. This is particularly critical as the `Compare()` function is used in multiple places across the codebase including incremental sync, backfill operations, and binlog processing.

---

#### **Proposed Solution**

Add comprehensive unit tests for the logic inside `compare.go`.  
This will help:

* Increase overall test coverage
* Improve comparison reliability across different data types
* Catch bugs earlier during CI
* Make contributions safer

For reference on how existing unit tests are structured, you can refer to **PR #273**.

---

#### **Notes for Contributors**

* Add tests covering most of the scenarios and edge cases which should indicate both pass and fail scenarios.
* Follow the existing test file structure and naming conventions to maintain consistency with other parts.

### PR DESCRIPTION
**Pull request by TheRootDaemon**
# Description

Fixes #650

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

```sh
 ~/prs/olake/utils/typeutils git:(compare.go) : go test -v -run TestCompare
=== RUN   TestCompare_Nil
--- PASS: TestCompare_Nil (0.00s)
=== RUN   TestCompare_SignedIntegers
--- PASS: TestCompare_SignedIntegers (0.00s)
=== RUN   TestCompare_UnsignedIntegers
--- PASS: TestCompare_UnsignedIntegers (0.00s)
=== RUN   TestCompare_Floats
--- PASS: TestCompare_Floats (0.00s)
=== RUN   TestCompare_FloatEdgeCases
--- PASS: TestCompare_FloatEdgeCases (0.00s)
=== RUN   TestCompare_Time
--- PASS: TestCompare_Time (0.00s)
=== RUN   TestCompare_Bool
--- PASS: TestCompare_Bool (0.00s)
=== RUN   TestCompare_Fallback
--- PASS: TestCompare_Fallback (0.00s)
PASS
ok  	github.com/datazip-inc/olake/utils/typeutils	0.031s
```

# Screenshots or Recordings

## Documentation

- [ ] Documentation Link: [link to README, olake.io/docs, or olake-docs]
- [x] N/A (bug fix, refactor, or test changes only)

## Related PR's (If Any):

### CODE REVIEW

**Comment by nayanj98** [comment]
@TheRootDaemon Can you please rebase your PR to staging from master.

File: utils/typeutils/compare_test.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+func TestCompare_Floats(t *testing.T) {
+	assert.Equal(t, 0, Compare(float64(3.3), float32(3.3)))
+	assert.Equal(t, -1, Compare(float32(1.1), float64(2.2)))
+	assert.Equal(t, 1, Compare(float64(5.5), float64(1.1)))
+}
+
+func TestCompare_FloatEdgeCases(t *testing.T) {
+	assert.Equal(t, 0, Compare(math.NaN(), math.NaN()))
+	assert.Equal(t, -1, Compare(math.NaN(), 1.0))
+	assert.Equal(t, 1, Compare(1.0, math.NaN()))
```
Comment (vaibhav-datazip) [inline utils/typeutils/compare_test.go]: please follow the pattern of unit test we are using for other files as well , the file name function name and that contains all the test cases which are then iterated by a for loop instead of making function for each datatype.

Comment (TheRootDaemon) [reply utils/typeutils/compare_test.go]: @vaibhav-datazip, Thanks for the review, I've refactored the tests. Please take a look.

Comment (vaibhav-datazip) [reply utils/typeutils/compare_test.go]: please rebase it to staging first, its currently raised on top of master branch. also, please avoid force pushing changes

Comment (TheRootDaemon) [reply utils/typeutils/compare_test.go]: Sorry, I thought I was supposed to do that locally and force push it in the remote repo.
I have changed the base to `staging`, I'll be careful from here on.

Thanks.

**Comment by nayanj98** [comment]
@TheRootDaemon Also once you are done with your work please add a video showing how you have tested your work, just makes it easier for us to know you are in the correct direction. And once the video is out I can assign someone to review your work

**Comment by TheRootDaemon** [comment]
> @TheRootDaemon Also once you are done with your work please add a video showing how you have tested your work, just makes it easier for us to know you are in the correct direction. And once the video is out I can assign someone to review your work

Ok sure.

**Comment by TheRootDaemon** [comment]
Hi @nayanj98, 

I have recorded a [loom](https://www.loom.com/share/20de14cf81004eb58b9a5919af575ff8) explaining my implementation and how I tested them. I hope this helps.

Thanks.

**Comment by Akshay-datazip** [comment]
> Hi @nayanj98,
> 
> I have recorded a [loom](https://www.loom.com/share/20de14cf81004eb58b9a5919af575ff8) explaining my implementation and how I tested them. I hope this helps.
> 
> Thanks.

Hi thanks for sharing this . team will get back on this soon

**Comment by TheRootDaemon** [comment]
> > Hi @nayanj98,
> > I have recorded a [loom](https://www.loom.com/share/20de14cf81004eb58b9a5919af575ff8) explaining my implementation and how I tested them. I hope this helps.
> > Thanks.
> 
> Hi thanks for sharing this . team will get back on this soon

Great, Thanks and I hope the video was clear and relevant.

**Comment by nayanj98** [comment]
@TheRootDaemon Assigning @ImDoubD-datazip to review your work

**Comment by badalprasadsingh** [comment]
@TheRootDaemon, check my comments on issue #719, and update the PR accordingly.

File: utils/typeutils/compare.go [hunk 2]
```diff
@@ -22,32 +23,45 @@ func Compare(a, b any) int {
 
 	switch aVal := a.(type) {
 	case uint, uint8, uint16, uint32, uint64:
-		aUint := reflect.ValueOf(a).Convert(reflect.TypeOf(uint64(0))).Uint()
-		bUint := reflect.ValueOf(b).Convert(reflect.TypeOf(uint64(0))).Uint()
+		aUint := reflect.ValueOf(a).Convert(reflect.TypeFor[uint64]()).Uint()
+		bUint := reflect.ValueOf(b).Convert(reflect.TypeFor[uint64]()).Uint()
```
Comment (badalprasadsingh) [inline utils/typeutils/compare.go]: why use `TypeFor` instead of `TypeOf`? `TypeFor` internally calls `TypeOf` if you check its code. 

```go
// TypeFor returns the [Type] that represents the type argument T.
func TypeFor[T any]() Type {
	var v T
	if t := TypeOf(v); t != nil {
		return t // optimize for T being a non-interface kind
	}
	return TypeOf((*T)(nil)).Elem() // only for an interface kind
}
```

File: utils/typeutils/compare.go [hunk 3]
```diff
@@ -22,32 +23,45 @@ func Compare(a, b any) int {
...
-		return 0
+
+		const eps = 1e-6
+		diff := aFloat - bFloat
+		if math.Abs(diff) < eps {
+			return 0
+		} else if diff < 0 {
+			return -1
+		}
+		return 1
```
Comment (badalprasadsingh) [inline utils/typeutils/compare.go]: This is the reason because of which you are able to differentiate a `float32(3.3)` value from a `float64(3.3)` value. Because it compares only upto fixed absolute epsilon (1e-6).

But, it creates a bug when you compare `float64` values with their difference is below the epsilon (1e-6). For example when you compare, `float64(3.99999997)` against `float64(3.99999998)` your logic will say that they are equal. This is a major bug. Remove this logic.

File: utils/typeutils/compare_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+		// floats
+		{"float_equal", float64(3.3), float32(3.3), 0},
+		{"float_less", float32(1.1), float64(2.2), -1},
+		{"float_greater", float64(5.5), float64(1.1), 1},
+
+		// float edge cases
+		{"nan_vs_nan", math.NaN(), math.NaN(), 0},
+		{"nan_vs_value", math.NaN(), 1.0, -1},
+		{"value_vs_nan", 1.0, math.NaN(), 1},
```
Comment (badalprasadsingh) [inline utils/typeutils/compare_test.go]: Getting a `math.Nan()` value in OLake's code architecture is very rare. Let's just remove the logic as well as the test for it.

File: utils/typeutils/compare.go [hunk 5]
```diff
@@ -22,32 +23,45 @@ func Compare(a, b any) int {
...
 		} else if aInt > bInt {
 			return 1
 		}
 		return 0
 	case float32, float64:
-		aFloat := reflect.ValueOf(a).Convert(reflect.TypeOf(float64(0))).Float()
-		bFloat := reflect.ValueOf(b).Convert(reflect.TypeOf(float64(0))).Float()
-		if aFloat < bFloat {
+		aFloat := reflect.ValueOf(a).Convert(reflect.TypeFor[float64]()).Float()
+		bFloat := reflect.ValueOf(b).Convert(reflect.TypeFor[float64]()).Float()
```
Comment (badalprasadsingh) [inline utils/typeutils/compare.go]: give a reason for having `TypeFor` instead of `TypeOf` or just remove `TypeFor` from everywhere you added.

File: utils/typeutils/compare.go (hunk 3 above)
Comment (badalprasadsingh) [inline utils/typeutils/compare.go]: This `return 1` at the end is the reason your `inf_vs_inf` case returns 1. This is not good way to handle this.

File: utils/typeutils/compare.go [hunk 6]
```diff
@@ -22,32 +23,45 @@ func Compare(a, b any) int {
...
+		if math.IsNaN(aFloat) {
+			if math.IsNaN(bFloat) {
+				return 0
+			}
 			return -1
-		} else if aFloat > bFloat {
+		}
+		if math.IsNaN(bFloat) {
 			return 1
 		}
```
Comment (badalprasadsingh) [inline utils/typeutils/compare.go]: Let's not handle the `math.NaN` case. In our current code architecture of OLake, we shouldn't be getting a case where we need to handle `math.NaN` case.

**Review by badalprasadsingh — CHANGES_REQUESTED** [review]

**Comment by TheRootDaemon** [comment]
I have updated the code, let me know if I missed something else.

File: utils/typeutils/compare_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		// floats
+		{"float_equal", float64(3.3), float64(3.3), 0},
+		{"float_less", float32(1.1), float32(2.2), -1},
+		{"float_greater", float64(5.5), float64(1.1), 1},
+
+		// time
+		{"time_equal", now, now, 0},
+		{"time_less", now, later, -1},
+		{"time_greater", later, now, 1},
+
```
Comment (badalprasadsingh) [inline utils/typeutils/compare_test.go:81]: We have a [Time struct in time.go](https://github.com/datazip-inc/olake/blob/8a05f172b547586c3215af6f335174e59e90b801/utils/typeutils/time.go#L8-L10), that is same as time.Time, but incase in future that changes the unit test should test that correctly as that is a different code path. Please add these tests as well

```go
{"custom_time_equal", Time{Time: now}, Time{Time: now}, 0},
{"custom_time_less", Time{Time: now}, Time{Time: later}, -1},
{"custom_time_greater", Time{Time: later}, Time{Time: now}, 1},
```

**Comment by TheRootDaemon** [comment]
> Just one more comment, @TheRootDaemon , and then we should be good to go.
> 
> To avoid a large number of AI-generated PRs, we ask first-time contributors to share a short video explaining how they tested their changes. Even for simple unit tests, having a basic understanding of the product flow helps in identifying real-world edge cases.
> 
> Could you please share a brief video of running OLake and demonstrating how you tested this PR?
> 
> Thanks.

I did record a [loom](https://www.loom.com/share/20de14cf81004eb58b9a5919af575ff8), but this explains how I tested the epsilon based comparisons.

Since this PR contains the unit tests for `compare.go`, I would like to confirm the expectations.
Do you want the video which explains the unit tests where I go through the test cases, or something else like where I run olake locally and explaining exatly which part does these unit tests help ?

Thanks.

**Comment by badalprasadsingh** [comment]
Hey, please add a video on "how you ran OLake locally and explaining exatly which part does these unit tests help". That would be great.

Kindly also solve [this comment](https://github.com/datazip-inc/olake/pull/720#discussion_r2678460479) of adding test cases for `Time` struct and then can move ahead.

**Comment by TheRootDaemon** [comment]
Hi @badalprasadsingh,

I finally recorded a [loom](https://www.loom.com/share/7e948995c13046bdb025d8642ea0662a) on how the unit tests help olake. I know that I did take a while for this. Let me know if you want anything else.

Thanks.

**Comment by nayanj98** [comment]
@TheRootDaemon I am assigning @ImDoubD-datazip for second review.

File: utils/typeutils/compare_test.go
Comment (ImDoubD-datazip) [inline utils/typeutils/compare_test.go:1]: please check if all these tests are covered or not. If not, then please add them. all these cases are required because it is a highly-used utility function in olake.

- Type Mismatch

| Test Name | a | b |
|-----------|---|---|
| time_vs_int | time.Now() | 123 |
| time_vs_string | time.Now() | "2024-01-01" |
| bool_vs_int | true | 1 |
| bool_vs_string | false | "false" |
| uint_vs_signed_negative | uint64(5) | int64(-1) |
| signed_negative_vs_uint | int64(-1) | uint64(5) |
| uint_vs_int_positive | uint64(100) | int64(100) |
| int_vs_uint | int64(50) | uint64(50) |
| stdlib_time_vs_custom_time | time.Now() | Time{Time: now} |



- Custom Time vs Default Handling

| Test Name | a | b |
|-----------|---|---|
| custom_time_vs_stdlib_time | Time{Time: now} | now |
| only_a_is_custom_time | Time{Time: now} | 123 |
| only_b_is_custom_time | "hello" | Time{Time: now} |



- Floating Point Edge Cases (NaN, Inf, Precision)

| Test Name | a | b |
|-----------|---|---|
| positive_inf_vs_positive_inf | math.Inf(1) | math.Inf(1) |
| negative_inf_vs_negative_inf | math.Inf(-1) | math.Inf(-1) |
| positive_inf_vs_negative_inf | math.Inf(1) | math.Inf(-1) |
| negative_inf_vs_positive_inf | math.Inf(-1) | math.Inf(1) |
| positive_inf_vs_number | math.Inf(1) | 1000000.0 |
| negative_inf_vs_number | math.Inf(-1) | -1000000.0 |
| zero_vs_negative_zero | 0.0 | -0.0 |
| float_precision_issue | 0.1 + 0.2 | 0.3 |




- Integer Boundary Values

| Test Name | a | b |
|-----------|---|---|
| int64_max_vs_max | math.MaxInt64 | math.MaxInt64 |
| int64_min_vs_min | math.MinInt64 | math.MinInt64 |
| int64_max_vs_min | math.MaxInt64 | math.MinInt64 |
| int64_min_vs_max | math.MinInt64 | math.MaxInt64 |
| int8_max_vs_min | int8(127) | int8(-128) |
| uint64_max_vs_max | uint64(math.MaxUint64) | uint64(math.MaxUint64) |
| uint64_max_vs_zero | uint64(math.MaxUint64) | uint64(0) |
| uint64_zero_vs_max | uint64(0) | uint64(math.MaxUint64) |



- Cross-Width Integer Comparisons

| Test Name | a | b |
|-----------|---|---|
| int8_vs_int64_equal | int8(100) | int64(100) |
| int64_vs_int8_less | int64(-1000) | int8(-1) |
| uint8_vs_uint64_equal | uint8(255) | uint64(255) |
| uint64_vs_uint8_greater | uint64(1000) | uint8(255) |
| int_vs_int32 | int(42) | int32(42) |
| uint_vs_uint32 | uint(42) | uint32(42) |



- Float32 vs Float64

| Test Name | a | b |
|-----------|---|---|
| float32_vs_float64_equal | float32(1.5) | float64(1.5) |
| float64_vs_float32_less | float64(1.0) | float32(2.0) |
| float32_vs_float64_greater | float32(3.0) | float64(1.0) |




- String & Fallback Comparisons

| Test Name | a | b |
|-----------|---|---|
| empty_string_vs_empty | "" | "" |
| empty_string_vs_nonempty | "" | "a" |
| nonempty_vs_empty_string | "a" | "" |
| numeric_string_lex_order | "10" | "9" |
| numeric_string_lex_order_2 | "2" | "10" |
| case_sensitive_comparison | "Apple" | "apple" |
| unicode_comparison | "α" | "β" |
| slice_fallback | []int{1, 2} | []int{1, 3} |
| map_fallback | map[string]int{"a": 1} | map[string]int{"a": 2} |




- Time Comparisons

| Test Name | a | b |
|-----------|---|---|
| time_zero_vs_zero | time.Time{} | time.Time{} |
| time_zero_vs_now | time.Time{} | time.Now() |
| time_nanosecond_diff | now | now.Add(time.Nanosecond) |
| time_same_instant_diff_tz | now.UTC() | now.In(loc) |
| custom_time_zero | Time{} | Time{} |




- Pointer, Interface, Error Fallbacks

| Test Name | a | b |
|-----------|---|---|
| pointer_to_int | &intVal | &intVal2 |
| interface_wrapped_int | interface{}(int(5)) | interface{}(int(5)) |
| error_type_fallback | errors.New("a") | errors.New("b") |

**Comment by ImDoubD-datazip** [comment]
Hi @TheRootDaemon , reviewed.

**Comment by TheRootDaemon** [comment]
Hi @ImDoubD-datazip,

I have added the missing test cases. Also I want to address a few things.

1. Comparing `math.NaN()` and the `float32 vs float64` values:

I had opened up an [issue](https://github.com/datazip-inc/olake/issues/719) regarding this. Please go through them.
I also had a snippet of code to handle math.NaN() values, right now it returns equal for everything when compared with `math.NaN`. I did remove them after discussing it with @badalprasadsingh.

```sh
##############
# Test Cases #
		{"nan_vs_nan", math.NaN(), math.NaN(), 0},
		{"nan_vs_value", math.NaN(), 1.0, -1},
		{"value_vs_nan", 1.0, math.NaN(), 1},
##############

 ~/prs/olake/utils/typeutils git:(compare.go ~) : ✗ go test -run TestCompare
--- FAIL: TestCompare (0.00s)
    --- FAIL: TestCompare/nan_vs_value (0.00s)
        compare_test.go:116:
            	Error Trace:	/home/therootdaemon/prs/olake/utils/typeutils/compare_test.go:116
            	Error:      	Not equal:
            	            	expected: -1
            	            	actual  : 0
            	Test:       	TestCompare/nan_vs_value
    --- FAIL: TestCompare/value_vs_nan (0.00s)
        compare_test.go:116:
            	Error Trace:	/home/therootdaemon/prs/olake/utils/typeutils/compare_test.go:116
            	Error:      	Not equal:
            	            	expected: 1
            	            	actual  : 0
            	Test:       	TestCompare/value_vs_nan
FAIL
exit status 1
FAIL	github.com/datazip-inc/olake/utils/typeutils	0.032s
```

Adding this snippet should handle the `math.NaN` edge cases.
```go
if math.IsNaN(aFloat) {
	if math.IsNaN(bFloat) {
		return 0
	}
	return -1
}
if math.IsNaN(bFloat) {
	return 1
}
```
2. Type Mismatches:

Right now the **code panics** when there are type mismatches. How do you want me to address them ??

---
I wanted to clarify these things before I add the test cases for these cases.

Thanks

**Comment by ImDoubD-datazip** [comment]
@TheRootDaemon , Hi no need to add comparison with Math.Nan. I had to remove it from the cases, but it slipped from my mind. Let me check with type mismatch once, will get back to you on this one.

File: utils/typeutils/compare.go [hunk 8]
```diff
@@ -49,7 +49,13 @@ func Compare(a, b any) int {
...
 		return 0
 	case time.Time:
-		bTime := b.(time.Time)
+		var bTime time.Time
+		if t, ok := b.(time.Time); ok {
+			bTime = t
+		} else if t, ok := b.(Time); ok {
+			bTime = t.Time
+		}
```
Comment (ImDoubD-datazip) [inline utils/typeutils/compare.go]: please don't change this. If you are still changing this, then do provide a case where your change might be required.

Comment (TheRootDaemon) [reply utils/typeutils/compare.go]: Hey as you mentioned when I added the test cases for the custom `Time` vs the stdlib's `Time.time` I noticed that the code paciked.

The changes that I made does not change the logic but allows it to compare the custom time cases.

I have added the test cases to test them too.
The existing code allowed me to compare  custom `Time` with `Time` and the standard `Time.time` with itself.

But now it can compare `Time` with `Time.time` with panicking
```sh
# {"custom_time_vs_stdlib_time_equal", Time{Time: now}, now, 0},
# {"custom_time_vs_stdlib_time_less",  Time{Time: now}, later, -1},
# {"custom_time_vs_stdlib_time_greater", later, Time{Time: now}, 1},

 ~/olake/utils/typeutils git:(compare.go ~) : go test -run TestCompare
--- FAIL: TestCompare (0.00s)
    --- FAIL: TestCompare/custom_time_vs_stdlib_time_greater (0.00s)
panic: interface conversion: interface {} is typeutils.Time, not time.Time [recovered, repanicked]
```

Comment (ImDoubD-datazip) [reply utils/typeutils/compare.go]: please remove this case

File: utils/typeutils/compare_test.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+	now := time.Now()
+	later := now.Add(time.Hour)
+	intval_1, intval_2 := 1, 1
+
+	testCases := []struct {
+		name     string
+		a        interface{}
+		b        interface{}
+		expected int
+	}{
```
Comment (ImDoubD-datazip) [inline utils/typeutils/compare_test.go]: also can you use proper naming for variables. a,b, intval1, intval2, later are not that appropriate

Comment (TheRootDaemon) [reply utils/typeutils/compare_test.go]: Done

File: utils/typeutils/compare.go (hunk 8 above)
Comment (TheRootDaemon) [reply utils/typeutils/compare.go]: I have removed those test cases, also I have reverted the code back to what it was previosly.

File: utils/typeutils/compare_test.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+		// unsigned integers
+		{"unsigned_int_equal", uint64(5), uint(5), 0},
+		{"unsigned_int_less", uint32(1), uint8(2), -1},
+		{"unsigned_int_greater", uint(8), uint16(1), 1},
+
+		// int8 boundaries
+		{"int8_max_vs_min", int8(127), int8(-128), 1},
+
+		// int64 boundaries
```
Comment (ImDoubD-datazip) [inline utils/typeutils/compare_test.go:46]: please add int32 and uint32 boundaries case.

**Comment by TheRootDaemon** [comment]
> also can you provide the video of tests after these changes.

I have added the test cases and the here is the [video](https://www.loom.com/share/7b48d4faae5548579bbc55b8e98bd0be) after the changes.

File: utils/typeutils/compare_test.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCompare(t *testing.T) {
+	currentTime := time.Now()
+	futureTime := currentTime.Add(time.Hour)
+	firstPointerTarget, secondPointerTarget := 1, 1
```
Comment (ImDoubD-datazip) [inline utils/typeutils/compare_test.go]: currentTime -> baseTime
futureTime -> laterTime
firstPointerTarget -> leftPointerTarget
secondPointerTarget  -> rightPointerTarget

**Comment by ImDoubD-datazip** [comment]
LGTM

**Comment by TheRootDaemon** [comment]
Since there are two other people who have reviewed the PR, should I get an approval from them ?

**Comment by ImDoubD-datazip** [comment]
Yes

**Comment by TheRootDaemon** [comment]
Sorry for that, I fixed the misspelled words which caused errors in the golangci-lint.

File: utils/typeutils/compare_test.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+		// fallback
+		{"fallback_string_vs_int", "123", 123, 0},
+		{"fallback_string_less", "apple", "banana", -1},
+		{"fallback_struct_greater", struct{ A int }{2}, struct{ A int }{1}, 1},
+		{"fallback_slice", []int{1, 2, 56}, []float32{2, 3}, -1},
+		{"fallback_map", map[string]int{"z": 1}, map[string]int{"b": 34}, 1},
+		{"fallback_errors", errors.New("a"), errors.New("A"), 1},
+		{"fallback_pointers_lesser", &leftPointerTarget, &rightPointerTarget, -1},
+		{"fallback_pointers_greater", &rightPointerTarget, &leftPointerTarget, 1},
```
Comment (ImDoubD-datazip) [inline utils/typeutils/compare_test.go]: i don't think there is sense in comparing pointer with greater and lesser format as relative order of Go allocating memory to these memory is non-deterministic. for now remove it.

Comment (TheRootDaemon) [reply utils/typeutils/compare_test.go]: @ImDoubD-datazip, Yeah I thought like that as well.

Since you mentioned that in the list of test cases that are missing, I added it anyways.
So would you like to have this test case instead ?

Since they both point at the same memory location, the assertion would make sense, since they are equal.

```go
pointerTarget := 1
leftPointerTarget, rightPointerTarget := &pointerTarget, &pointerTarget

// {"fallback_pointers", leftPointerTarget, rightPointerTarget, 0},
```

Also removing those two test cases would mean that the variables `leftPointerTarget`, `rightPointerTarget` will remain unused, which would mean that no test cases for pointers.

Shall I add the same pointer assertion or shall I remove it ?

Comment (ImDoubD-datazip) [reply utils/typeutils/compare_test.go]: @TheRootDaemon  no need for this. this is quite understood. you can remove the pointers test and associated variables as well.

Comment (TheRootDaemon) [reply utils/typeutils/compare_test.go]: @ImDoubD-datazip, I have removed the test cases and variables related to that in the latest commit [(acc617e)](https://github.com/datazip-inc/olake/pull/720/changes/acc617ee47a01298f4dbfa753e9bf119b753e963)

**Comment by ImDoubD-datazip** [comment]
LGTM
